File: code-recipes/java/src/main/java/vendorretailprocurement/orders/SubmitAcknowledgementRecipe.java
package vendorretailprocurement.orders;

import com.amazon.SellingPartnerAPIAA.LWAException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;
import org.threeten.bp.OffsetDateTime;
import software.amazon.spapi.ApiException;
import software.amazon.spapi.api.vendor.orders.v1.VendorOrdersApi;
import software.amazon.spapi.models.vendor.orders.v1.*;
import util.Constants;
import util.Recipe;

import java.util.List;

/**
 * Vendor Orders API Recipe: Submit Acknowledgement
 * =================================================
 * 
 * This recipe demonstrates how to submit acknowledgements for purchase orders
 * using the Vendor Orders API. Vendors use this to accept or reject orders.
 * 
 * Use cases:
 * - Accept a purchase order fully
 * - Partially accept an order (accept some items, reject others)
 * - Reject an order with a reason code
 * 
 * API Operations used:
 * - submitAcknowledgement: Submits acknowledgements for one or more purchase orders
 */
public class SubmitAcknowledgementRecipe extends Recipe {

    private static final Logger logger = LoggerFactory.getLogger(SubmitAcknowledgementRecipe.class);

    private final VendorOrdersApi vendorOrdersApi = new VendorOrdersApi.Builder()
            .lwaAuthorizationCredentials(lwaCredentials)
            .endpoint(Constants.BACKEND_URL)
            .build();

    @Override
    public void start() {
        String transactionId = submitOrderAcknowledgement(
                "TestOrder202", "API01", "028877454078", 10, 10, "10.2"
        );
        logger.info("Acknowledgement submitted. Transaction ID: {}", transactionId);
    }

    /**
     * Submit an acknowledgement to accept a purchase order.
     */
    public String submitOrderAcknowledgement(
            String purchaseOrderNumber,
            String sellingPartyId,
            String vendorProductId,
            int orderedQuantity,
            int acceptedQuantity,
            String netCostAmount) {
        try {
            OrderItemAcknowledgement itemAck = new OrderItemAcknowledgement()
                    .acknowledgementCode(OrderItemAcknowledgement.AcknowledgementCodeEnum.ACCEPTED)
                    .acknowledgedQuantity(new ItemQuantity().amount(acceptedQuantity));

            OrderAcknowledgementItem item = new OrderAcknowledgementItem()
                    .vendorProductIdentifier(vendorProductId)
                    .orderedQuantity(new ItemQuantity().amount(orderedQuantity))
                    .netCost(new Money().amount(netCostAmount))
                    .itemAcknowledgements(List.of(itemAck));

            OrderAcknowledgement orderAck = new OrderAcknowledgement()
                    .purchaseOrderNumber(purchaseOrderNumber)
                    .sellingParty(new PartyIdentification().partyId(sellingPartyId))
                    .acknowledgementDate(OffsetDateTime.now())
                    .items(List.of(item));

            SubmitAcknowledgementRequest request = new SubmitAcknowledgementRequest()
                    .acknowledgements(List.of(orderAck));

            SubmitAcknowledgementResponse response = vendorOrdersApi.submitAcknowledgement(request);

            if (response.getPayload() != null) {
                logger.info("Successfully submitted acknowledgement for PO: {}", purchaseOrderNumber);
                return response.getPayload().getTransactionId();
            }
            logger.warn("No transaction ID returned for PO: {}", purchaseOrderNumber);
            return null;
        } catch (ApiException | LWAException e) {
            logger.error("Error submitting acknowledgement for PO {}: {}", purchaseOrderNumber, e.getMessage(), e);
            throw new RuntimeException("Failed to submit acknowledgement", e);
        }
    }

    /**
     * Submit a rejection for a purchase order item.
     */
    public String submitOrderRejection(
            String purchaseOrderNumber,
            String sellingPartyId,
            String vendorProductId,
            int orderedQuantity,
            OrderItemAcknowledgement.RejectionReasonEnum rejectionReason) {
        try {
            OrderItemAcknowledgement itemAck = new OrderItemAcknowledgement()
                    .acknowledgementCode(OrderItemAcknowledgement.AcknowledgementCodeEnum.REJECTED)
                    .rejectionReason(rejectionReason)
                    .acknowledgedQuantity(new ItemQuantity().amount(0));

            OrderAcknowledgementItem item = new OrderAcknowledgementItem()
                    .vendorProductIdentifier(vendorProductId)
                    .orderedQuantity(new ItemQuantity().amount(orderedQuantity))
                    .itemAcknowledgements(List.of(itemAck));

            OrderAcknowledgement orderAck = new OrderAcknowledgement()
                    .purchaseOrderNumber(purchaseOrderNumber)
                    .sellingParty(new PartyIdentification().partyId(sellingPartyId))
                    .acknowledgementDate(OffsetDateTime.now())
                    .items(List.of(item));

            SubmitAcknowledgementRequest request = new SubmitAcknowledgementRequest()
                    .acknowledgements(List.of(orderAck));

            SubmitAcknowledgementResponse response = vendorOrdersApi.submitAcknowledgement(request);

            if (response.getPayload() != null) {
                logger.info("Successfully submitted rejection for PO: {}", purchaseOrderNumber);
                return response.getPayload().getTransactionId();
            }
            return null;
        } catch (ApiException | LWAException e) {
            logger.error("Error submitting rejection: {}", e.getMessage(), e);
            throw new RuntimeException("Failed to submit rejection", e);
        }
    }
}

File: code-recipes/python/src/recipes/vendor_retail_procurement/orders/submit_acknowledgement_recipe.py
"""
Vendor Orders API: Submit Acknowledgement
==========================================

Submits acknowledgements for purchase orders using the Vendor Orders API.
Vendors use this to accept or reject orders.

Use cases:
- Accept a purchase order fully
- Partially accept an order (accept some items, reject others)
- Reject an order with a reason code

API operation: submitAcknowledgement
"""

import logging
from datetime import datetime, timezone
from typing import Any, Dict, Optional

from spapi import SPAPIClient, SPAPIConfig
from spapi.api.vendor_orders_v1.vendor_orders_api import VendorOrdersApi

from src import config
from src.util.recipe import Recipe

logger = logging.getLogger(__name__)


class SubmitAcknowledgementRecipe(Recipe):

    def __init__(
        self,
        sp_config: Optional[SPAPIConfig] = None,
        vendor_orders_api: Optional[VendorOrdersApi] = None,
    ) -> None:
        super().__init__(config=sp_config)
        self._vendor_orders_api = vendor_orders_api

    @property
    def vendor_orders_api(self) -> VendorOrdersApi:
        if self._vendor_orders_api is None:
            client = SPAPIClient(
                self.config,
                oauth_endpoint=f"{config.backend_url}/auth/o2/token",
                endpoint=config.backend_url,
            )
            self._vendor_orders_api = VendorOrdersApi(client.api_client)
            logger.info("VendorOrdersApi client initialized")
        return self._vendor_orders_api

    def submit_order_acknowledgement(
        self,
        purchase_order_number: str,
        selling_party_id: str,
        vendor_product_id: str,
        ordered_quantity: int,
        accepted_quantity: int,
        net_cost_amount: str,
    ) -> Optional[str]:
        """Submit an acknowledgement to accept a purchase order."""
        logger.info("Submitting acknowledgement for PO: %s", purchase_order_number)
        body = {
            "acknowledgements": [
                {
                    "purchaseOrderNumber": purchase_order_number,
                    "sellingParty": {"partyId": selling_party_id},
                    "acknowledgementDate": datetime.now(timezone.utc).isoformat(),
                    "items": [
                        {
                            "vendorProductIdentifier": vendor_product_id,
                            "orderedQuantity": {"amount": ordered_quantity},
                            "netCost": {"amount": net_cost_amount},
                            "itemAcknowledgements": [
                                {
                                    "acknowledgementCode": "Accepted",
                                    "acknowledgedQuantity": {"amount": accepted_quantity},
                                }
                            ],
                        }
                    ],
                }
            ]
        }
        response = self.vendor_orders_api.submit_acknowledgement(body=body)
        transaction_id = self._extract_transaction_id(response)
        if transaction_id:
            logger.info("Successfully submitted acknowledgement for PO: %s", purchase_order_number)
            logger.info("Transaction ID: %s", transaction_id)
        else:
            logger.warning("No transaction ID returned for PO: %s", purchase_order_number)
        return transaction_id

    def submit_order_rejection(
        self,
        purchase_order_number: str,
        selling_party_id: str,
        vendor_product_id: str,
        ordered_quantity: int,
        rejection_reason: str = "TemporarilyUnavailable",
    ) -> Optional[str]:
        """Submit a rejection for a purchase order item."""
        logger.info("Submitting rejection for PO: %s", purchase_order_number)
        body = {
            "acknowledgements": [
                {
                    "purchaseOrderNumber": purchase_order_number,
                    "sellingParty": {"partyId": selling_party_id},
                    "acknowledgementDate": datetime.now(timezone.utc).isoformat(),
                    "items": [
                        {
                            "vendorProductIdentifier": vendor_product_id,
                            "orderedQuantity": {"amount": ordered_quantity},
                            "itemAcknowledgements": [
                                {
                                    "acknowledgementCode": "Rejected",
                                    "rejectionReason": rejection_reason,
                                    "acknowledgedQuantity": {"amount": 0},
                                }
                            ],
                        }
                    ],
                }
            ]
        }
        response = self.vendor_orders_api.submit_acknowledgement(body=body)
        transaction_id = self._extract_transaction_id(response)
        if transaction_id:
            logger.info("Successfully submitted rejection for PO: %s", purchase_order_number)
        return transaction_id

    def start(self) -> None:
        transaction_id = self.submit_order_acknowledgement(
            purchase_order_number="TestOrder202",
            selling_party_id="API01",
            vendor_product_id="028877454078",
            ordered_quantity=10,
            accepted_quantity=10,
            net_cost_amount="10.2",
        )
        logger.info("Acknowledgement submitted. Transaction ID: %s", transaction_id)

    @staticmethod
    def _extract_transaction_id(response: Any) -> Optional[str]:
        if hasattr(response, "to_dict"):
            response = response.to_dict()
        payload = response.get("payload") or {}
        return payload.get("transaction_id")


if __name__ == "__main__":
    logging.basicConfig(level=logging.INFO)
    recipe = SubmitAcknowledgementRecipe()
    recipe.start()
